Add changes for implementing separate worker nodes

rlkennedyreid opened the pull request:
- Adding python3 as a yum install, as python3 binary is needed for slurm build, but user-scoped python version is more reliable for using in slurm jobs
- Added pyenv install steps to the slurmrestd user in their HOME directory
- Edited entrypoint to recognise an env variable for the node type
- Changed slurm conf file to expect a worker node
- TODO: work out a new shared space setup for the worker nodes and the host node
    - They can't be the current home directory, because mounting on separate nodes causes conflicts due to the files aready present in the image (dotfiles, .pyenv directory...)
 - Some build flags added to slurm that according to the docs should speed up one-time builds
 - Also added a user-env related flag to configure that appeared to desired behaviour

- Changes have been tested using commands on host node of the form `sbatch --uid=1002 --gid=1002 --export=NONE script`
- **Note:** Of most importance seems to be this `--export` flag and the script having the shebang `#!/bin/bash -l`
    - All other flags, even `--get-user-env` flag, seem to have no effect

rlkennedyreid committed:
Add changes for implementing separate worker nodes

- Adding python3 as a yum install, as python3 binary is needed for slurm build, but user-scoped python version is more reliable for using in slurm jobs
- Added pyenv install steps to the slurmrestd user in their HOME directory
- Edited entrypoint to recognise an env variable for the node type
- Changed slurm conf file to expect a worker node
- TODO: work out a new shared space setup for the worker nodes and the host node

drewsilcock commented on Dockerfile:
```diff
@@ -55,43 +55,26 @@ RUN set -ex \
     && git clone -b ${SLURM_TAG} --single-branch --depth=1 https://github.com/SchedMD/slurm.git \
     && pushd slurm \
-    && ./configure --prefix=/usr --sysconfdir=/etc/slurm \
+    && ./configure --quiet --disable-dependency-tracking --prefix=/usr --sysconfdir=/etc/slurm --enable-load-env-no-login \
```
What do these options do?

rlkennedyreid replied:
Some relevant parts of Slurm repo I could find

`--enable-load-env-no-login`:
[x_ac_env.m4](https://github.com/SchedMD/slurm/blob/4801c60da4784346c4bf830a0b198364012e44ee/auxdir/x_ac_env.m4#L18)
[config.h.in](https://github.com/SchedMD/slurm/blob/1ca5f08f038c389b5073ea3f00de3b4aef628467/config.h.in#L416)
[env.c](https://github.com/SchedMD/slurm/blob/1225a45876c055b75654e2817571bb8393fc414b/src/common/env.c#L2065)

So I got the impression that this allowed loading of user environments without a `.login` file being present. I thought this may have been the source of issues because bash doesn't use this dotfile, and centos doesn't have it in user directories by default. I don't think adding it really affected anything though. Very hard to trial and error because of the need to do a full rebuild.

`--disable-dependency-tracking`:
[configure](https://github.com/SchedMD/slurm/blob/f4bfd747434a92715fa1e5f996ea498532450b0a/configure#L1693)

So I got the impression this would speed up the build in a docker build - but not quite sure what the implications are.

If you think best to remove, I'm happy with that. Will attempt to rebuild and check the affects!

drewsilcock replied:
Based on the `env.c` this is the different it seems to make:

https://superuser.com/questions/453988/whats-the-difference-between-su-with-and-without-hyphen

Not sure what implications this has for the rest of Slurm - if it isn't required to make the environment work maybe best to leave it out as we don't fully understand what it does.

rlkennedyreid replied:
Ah yeah I thought the same - though I think I misinterpreted which one brought the calling environment in. It does seem to be we want to behavior **with** the hyphen, so I've remove the flag now.

Interestingly - this does seem to be an issue/bug with slurm. The `--export=NONE` is needed to stop environment variables coming through from the calling environment.

rlkennedyreid replied:
Also removed the depency tracking flag

drewsilcock commented on docker-compose.yml:
```diff
@@ -1,27 +1,80 @@
+    slurmctl:
+        build: .
+        hostname: slurmctl
+        command: "controller"
```
Doesn't look like this command does anything, it's the environment variable that's used. Is that right?

rlkennedyreid replied:
Correct, yeah. This is from a previous iteration where I used the command rather than an env. But this meant you had to hardcode `exec bash` rather than `exec $@` - so went for env.

Pushed the fix now

drewsilcock reviewed (commented):
Looks good, just left a few questions on some specific bits.

rlkennedyreid committed:
Remove unused commands from docker-compose

rlkennedyreid committed:
Add deltarpm package before update steps to decrease download and install time

Also neatun up configure flags in slurm build

rlkennedyreid committed:
Use file input to specify yum configuration and upgrade rather than update

Upgrade removes obselete packages. File configuration allows disabling cache and making non-interactive without repeated flags

rlkennedyreid committed:
Remove unecessary echo check

rlkennedyreid commented:
Think that's all comments addressed. I made some minor changes to the yum install sections. It's now similar to the Dakota image - deltarpm to speed up upgrades, upgrade instead of update to remove obselete packages in the image, and using `yum.conf` rather than flags.

drewsilcock commented on Dockerfile:
```diff
@@ -9,11 +9,13 @@ LABEL org.opencontainers.image.source="https://github.com/drewsilcock/docker-cen
-    && yum -y update \
-    && yum -y install epel-release \
-    && yum -y install \
+    && cat input-yum.conf >> /etc/yum.conf \
```
Can also presumably do `COPY ./input-yaml.conf /etc` but doesn't really matter.
